documentation/Functional-tests.md

```diff
@@ -0,0 +1,25 @@
+# Functional test suite
+Functional tests are located in the folder `functional-tests`.
+
+To run functional tests, make sure you have Selenium and Seleniumbase installed (from backend requirements). Also, create an .ini file in the `functional-tests` folder, which should have the following setup:
+```
+[credentials]
+    username = {your_username}
+    password = {your_password}
+
+[corpus]
+    name = {name_of_corpus_your_user_can_access}
+    field = {multiple_choice_filter_field_name}
+    first_option = {first_option_of_that_field}
+    query = {query_term}
+```
+
+Note that in order for all tests to work, the corpus should also have associated *word models*
```

Comment:
This is not a full review, just a comment on a specific point.

This seems highly specific. Especially the last one: a developer who does not work on word models should not have to download word models to run functional tests.

My intuition would be that the current setup of tests is best understood as a test of whether the active environment is "valid" - if there are no word models, nothing word-modelly needs to be tested.

The more generalised version of that is to not include any of this information; instead, let the functional tests explore the interface to assemble a list of available corpora, then open the link of each corpus to see if it works, check if there is a link to word models, see if that works, etc. That is more work than including a lot of assumptions though.

Reply:
Agreed, the word model tests could also just be skipped if they're not present.

As for the assumptions - I think this is something we cannot completely abstract away. For instance, if we trigger the first filter of the first corpus, what values do we expect in the route? I'm also still struggling with this, but in essence, from what I understand, functional tests are built against a *running* application, and not against a sandboxed environment. We discussed the latter for GitHub actions, but I think both routes have merit. I believe if we filled in some of these "assumptions" for one of the live servers, we could get much better diagnostics of whether the server is behaving as it should.

Reply:
Hm, fair enough, these assumptions do make things easier. Still, I think you should be able to configure that features like word models are not present in your environment.

>  Then again, I don't agree that a "valid" setup is what we want to test here - it's that every part of the application reacts as it should. So that if we change one component, we can catch consequences to unrelated components, such as the word models, which may otherwise remain undetected.

> functional tests are built against a running application, and not against a sandboxed environment

This sounds contradictory to me, at least for an ostensibly modular / configurable application like I-analyzer. You can test whether a running environment is working as it should, or you can setup a sandbox that includes every module you want to test.

In the case of the first, you should be able to validate an environment that opts out of some I-analyzer features (such as not having word models or not having multiple choice filters), since that is still a working environment - and yes, that environment will still pass all tests if the word models are broken, because what you are testing is if *this* environment works as it should.

In the case of the second, you can actually make sure that any module is present and tested because you control the whole environment - but with obvious limitations because you are not testing a "real" environment.

I'm somewhat worried that making strong demands like this in the functional tests will make it harder to maintain I-analyzer as a modular application. This would especially get problematic if we would use a similar setup for things like each filter type, term frequency normalisation, document context, tagging, NER, etc.